engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java

```diff
@@ -1718,16 +1569,52 @@ def translateComment(comment: AST): Comment = {
   }
 
   private Option<IdentifiedLocation> getIdentifiedLocation(Tree ast) {
-    if (ast == null) {
-      return Option.empty();
+    return Option.apply(ast).map(ast_ -> {
+      int begin = Math.toIntExact(ast_.getStartCode());
+      int end = Math.toIntExact(ast_.getEndCode());
+      return new IdentifiedLocation(new Location(begin, end), Option.empty());
+    });
+  }
+
+  private Option<IdentifiedLocation> getIdentifiedLocation(ArgumentDefinition ast) {
+    // Note: In the IR, `DefinitionArgument` is a type of AST node; in the parser, it is not an AST-level type, but a
+    // type used only in specific locations. As a result, the IR expects it to have a source-code location, but the
+    // parser doesn't have a span for it. Here we synthesize one. This will not be necessary if we refactor IR so that
+    // `ArgumentDefinition` is not an AST node, though that change would have effects throughout the compiler and may
+    // not be worthwhile if IR is expected to be replaced.
+    long begin;
+    if (ast.getOpen() != null) {
+      begin = ast.getOpen().getStartCode();
+    } else if (ast.getOpen2() != null) {
+      begin = ast.getOpen2().getStartCode();
+    } else if (ast.getSuspension() != null) {
+      begin = ast.getSuspension().getStartCode();
+    } else {
+      begin = ast.getPattern().getStartCode();
     }
-    int begin, len;
-    {
-      begin = Math.toIntExact(ast.getStartCode());
-      int end = Math.toIntExact(ast.getEndCode());
-      len = end - begin;
+    int begin_ = Math.toIntExact(begin);
+    long end;
+    if (ast.getClose() != null) {
+      end = ast.getClose().getEndCode();
+    } else if (ast.getDefault() != null) {
+      end = ast.getDefault().getEquals().getEndCode();
+    } else if (ast.getClose2() != null) {
+      end = ast.getClose2().getEndCode();
+    } else if (ast.getType() != null) {
+      end = ast.getType().getOperator().getEndCode();
+    } else {
+      end = ast.getPattern().getEndCode();
     }
-    return Option.apply(new IdentifiedLocation(new Location(begin, begin + len), Option.empty()));
+    int end_ = Math.toIntExact(end);
+    return Option.apply(new IdentifiedLocation(new Location(begin_, end_), Option.empty()));
+  }
+
+  private Option<IdentifiedLocation> getIdentifiedLocation(Token ast) {
+    return Option.apply(ast).map(ast_ -> {
```

Comment:
There is a lot of _functional style_ rewrites in this file! @kazcw, looks like you can (unlike me) write functionally (even) in Java.